### PR TITLE
refactor: Unify lifecycle event handling with clearer attach/detach semantics

### PR DESCRIPTION
**Pull request by RicoSuter**
Lifecycle event handling previously conflated context-level attachment with property-level reference tracking, using ambiguous method names and internal flag enums that obscured intent.

## Breaking Changes

**API Renames:**
- `ILifecycleHandler.OnLifecycleEvent()` → `HandleLifecycleChange()` 
- `ILifecycleInterceptor.AttachTo/DetachFrom()` → `AttachSubjectToContext/DetachSubjectFromContext()`
- Event `SubjectDetached` → `SubjectDetaching` (fires BEFORE handlers for graph inspection)

**Flag Changes in `SubjectLifecycleChange`:**
- Internal `LifecycleEventFlags` enum replaced with explicit properties
- `IsAttached` → `IsContextAttach`
- `IsReferenceAdded` → `IsPropertyReferenceAdded`
- `IsReferenceRemoved` → `IsPropertyReferenceRemoved`
- `IsDetached` → `IsContextDetach`

## Architecture

**Split monolithic attach/detach into 4 focused methods:**
- `AttachToContext()` - Root subject enters graph (no property reference)
- `AttachToProperty()` - Subject attached via property reference  
- `DetachFromContext()` - Root subject leaves graph
- `DetachFromProperty()` - Subject detached from property reference

**Event ordering now symmetric and documented:**
- Attach: handlers → `SubjectAttached` event → property lifecycle
- Detach: property lifecycle → `SubjectDetaching` event → handlers

## Testing & Documentation

- Added `TestLifecycleHandler` with snapshot testing support
- 436 lines of graph behavior tests: `HierarchyTests`, `DagTests`, `CycleTests`
- Documents edge cases (cycles, DAGs, self-references) with verified snapshots
- New `subject-guidelines.md` covering lifecycle patterns

### CODE REVIEW

**Comment by RicoSuter** [comment]
@copilot please review this PR in-depth (correctness, does it make sense for a generic library, docs, tests, coverage, arch, software design, etc.) also give me a title and description

File: src/Namotion.Interceptor.Tracking.Tests/Lifecycle/LifecycleEventsTests.cs [hunk 1]
```diff
@@ -34,18 +34,19 @@ public void SubjectAttached_FiresWithCorrectReferenceCount()
...
     [Fact]
-    public void SubjectDetached_FiresWithCorrectReferenceCount()
+    public void SubjectDetaching_FiresWithCorrectReferenceCount()
     {
         // Arrange
         var context = InterceptorSubjectContext
             .Create()
-            .WithLifecycle();
+            .WithLifecycle()
+            .WithContextInheritance();
```
Comment (RicoSuter) [inline src/Namotion.Interceptor.Tracking.Tests/Lifecycle/LifecycleEventsTests.cs]: Problem that these are needed? Why?
